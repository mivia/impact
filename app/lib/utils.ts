import type { Product } from "../components/product/product";

export const getUniqueCartItems = (product: Product, cartItems: Product[]) => {
  let newCartItems = [];
  const productIndex = cartItems.findIndex((item: Product) => item.id === product.id)
  const isProductInTheCart = productIndex > -1

  if (!isProductInTheCart) {
    newCartItems = [...cartItems]
    newCartItems.push({ ...product })
  } else {
    newCartItems = [...cartItems]
    newCartItems[productIndex] = { ...product, quantity: cartItems[productIndex].quantity + 1 }
  }

  return newCartItems;
}

export class LocalStorageMock {
  store: Record<string, string>
  length: number

  constructor() {
    this.store = {};
    this.length = length;
  }

  clear() {
    this.store = {};
  }

  getItem(key: string) {
    return this.store[key] || null;
  }

  setItem(key: string, value: string) {
    this.store[key] = String(value);
  }

  removeItem(key: string) {
    delete this.store[key];
  }
}