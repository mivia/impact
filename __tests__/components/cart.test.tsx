import { render } from '@testing-library/react'

import Cart from '@/app/components/cart/cart'

import { LocalStorageMock } from '../../app/lib/utils';

// @ts-expect-error Don't have enough time to find a way of implementing the key property to above mock
global.localStorage = new LocalStorageMock;
 
it('renders Cart with a circle containing items count', () => {
  localStorage.setItem('cart', JSON.stringify([{id: '1', quantity: 1}]))
  const { container } = render(<Cart />)
  expect(container).toMatchSnapshot()
})

it('renders Cart without a circle containing items count', () => {
  localStorage.setItem('cart', JSON.stringify([]))
  const { container } = render(<Cart />)
  expect(container).toMatchSnapshot()
})