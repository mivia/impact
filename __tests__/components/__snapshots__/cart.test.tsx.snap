// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Cart with a circle containing items count 1`] = `
<div>
  <div
    class="cartContainer"
  >
    <a
      href="/cart"
    >
      <img
        alt="Basket icon"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="/basket-icon.svg"
        style="color: transparent;"
        width="50"
      />
      <div
        class="productsCountCircle"
      >
        <b>
          1
        </b>
      </div>
    </a>
  </div>
</div>
`;

exports[`renders Cart without a circle containing items count 1`] = `
<div>
  <div
    class="cartContainer"
  >
    <a
      href="/cart"
    >
      <img
        alt="Basket icon"
        data-nimg="1"
        decoding="async"
        height="50"
        loading="lazy"
        src="/basket-icon.svg"
        style="color: transparent;"
        width="50"
      />
    </a>
  </div>
</div>
`;
